plugin/query_predicates.lua

```diff
@@ -45,29 +45,37 @@ query.add_directive('set-lang-from-mimetype!', function(match, _, bufnr, pred, m
   end
 end, true)
 
-local injection_aliases = {
-  ex = 'elixir',
-  pl = 'perl',
-  sh = 'bash',
-  uxn = 'uxntal',
-  ts = 'typescript',
-}
-
+--- Set injection.language metadata from node text
+---
+--- If text does not yield an installed parser name, treat as `filetype`
+--- and look up corresponding parser.
+---
+--- Custom aliases can be added via |vim.treesitter.language.register()|.
 ---@param match (TSNode|nil)[]
 ---@param _ string
 ---@param bufnr integer
 ---@param pred string[]
 ---@return boolean|nil
-query.add_directive('set-lang-from-info-string!', function(match, _, bufnr, pred, metadata)
-  local capture_id = pred[2]
-  local node = match[capture_id]
+query.add_directive('inject-language!', function(match, _, bufnr, pred, metadata)
+  local node = match[pred[2]]
   if not node then
     return
   end
 
-  local injection_alias = vim.treesitter.get_node_text(node, bufnr)
-  local filetype = vim.filetype.match({ filename = 'a.' .. injection_alias })
-  metadata['injection.language'] = filetype or injection_aliases[injection_alias] or injection_alias
+  local has_parser = function(lang)
+    return vim._ts_has_language(lang)
+      or #vim.api.nvim_get_runtime_file('parser/' .. lang .. '.*', false) > 0
+  end
+
+  local alias = vim.treesitter.get_node_text(node, bufnr)
```

Comment:
It would be good if this somehow also respects offset.

No biggie otherwise.

Reply:
Is this relevant here? (Not opposed; just trying to understand the situation, so I can think how to implement it.)

Reply:
`offset` and maybe also `gsub`, if possible?

as sometimes the node text for `injection.language` can contain `"`, `'` or ` `, offset can be used to remove those unnecessary characters.

Reply:
I think it's relevant anywhere that `get_node_text()` is called in the context of queries.

Reply:
Well, right now this is specific to the markdown parser and not used anywhere else (although it might be?)

Or do I misunderstand the proposal? Could you give a specific example of what should be done here?

Reply:
Yes it should, that's why it has a `metadata` argument. I think that's all we need to add here.